### PR TITLE
[ADP-3479] Add addresses codec to deposit wallet

### PR DESCRIPTION
**Pull request by paolino**
- Add decoding and encoding between ledger addresses and bech32 or base58 under a network constraint to deposit wallet lib

ADP-3479

### CODE REVIEW

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+instance MonadFail CatchFail where
+    fail = CatchFail . Left
+
+ledgerAddressFlavor :: SL.Addr c -> AddressFlavor ()
+ledgerAddressFlavor (SL.AddrBootstrap _) = Bootstrap ()
+ledgerAddressFlavor _ = Shelley ()
+
+ledgerAddressNetwworkTag :: SL.Addr c -> NetworkTag
+ledgerAddressNetwworkTag addr = case SL.getNetwork addr of
+    SL.Testnet -> TestnetTag
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs:141]: we are sure we want isomorphic and so similar type to the ledger here?

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    SL.Mainnet -> MainnetTag
+
+inspectAddress
+    :: NetworkTag
+    -> AddressFlavor ByteString
+    -> Either DecodingError (AddressFlavor Address)
+inspectAddress nt a = inspect $ addressFlavor a
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs]: no chance and reason to use functionality from cardano-addresses here?

**Review by paweljakubas — APPROVED** [review]
nice PR! I left some remarks/questions. Also, what about adding explicit testing for inspectAddress?

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    bytes = B8.fromShort $ toShortByteString addr
+    bech32 = Bech32.encodeLenient hrp (dataPartFromBytes bytes)
+    hrp = humanPart tag
+encodeFlavoredAddress _ (Bootstrap addr) =
+    T.decodeUtf8 . encodeBase58 bitcoinAlphabet
+        $ B8.fromShort
+        $ toShortByteString addr
+
+-- | Encode an 'Address' to a 'Text' representation given the 'NetworkTag'.
+encodeAddress :: NetworkTag -> Address -> Text
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs]: you are big optimist here ... no chance to fail here? I would argue yes

**Review by paweljakubas — CHANGES_REQUESTED** [review]
Please add `EncodingError` and use it in encodeAddress

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs (hunk 1 above)
Comment (paolino) [reply lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs:141]: I do not have control on this, but happy to remove this when it's removed from the deposit agda code.

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs (hunk 2 above)
Comment (paolino) [reply lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs]: I got lost there and it seems to work badly with the network constraining aspect, but open to switch

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs (hunk 3 above)
Comment (paolino) [reply lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs]: yes network can be wrong

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs]: please see https://github.com/IntersectMBO/cardano-addresses/blob/master/core/lib/Cardano/Address/Style/Shelley.hs#L569

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs:141]: see also https://github.com/IntersectMBO/cardano-addresses/blob/master/core/lib/Cardano/Address/Style/Shelley.hs#L1085

**Review by paweljakubas — APPROVED** [review]
lgtm

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs (hunk 3 above)
Comment (paolino) [reply lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs]: We removed the network constraints

**Pull request merged**